### PR TITLE
Rework trie code

### PR DESCRIPTION
**Pull request by lightclient**
Rework the trie code to be more readable. Also pass the canonical trie tests. I'm a bit unsure about the typing, it feels like in some places `rlp.RLP` is less prescriptive than I could've been. But I'm not sure.

### CODE REVIEW

File: tests/helpers.py [hunk 1]
```diff
@@ -10,6 +10,14 @@
 )
 
 
+def sanitize(x: str) -> Bytes:
```
Comment (Bhargavasomu) [inline tests/helpers.py]: Should we name this `to_bytes`, and we can probably mention about the sanitization in the docstrings

File: src/eth1spec/trie.py [hunk 2]
```diff
@@ -13,206 +13,182 @@
...
     -------
-    encoded : `bytearray`
-        TODO
+    compressed : `bytearray`
+        Compact byte array.
     """
-    if verbose:
-        print("HP(", x, t, ")")
-    # x = bytes([int(d,16) for d in x.hex()])
     encoded = bytearray()
```
Comment (Bhargavasomu) [inline src/eth1spec/trie.py]: Shouldn't we change `encoded` to `compact` / `compressed` which is inline with the function name?

File: src/eth1spec/trie.py [hunk 3]
```diff
@@ -13,206 +13,182 @@
...
 
+    for preimage in obj:
+        # "secure" tries hash keys once before construction
+        key = crypto.keccak256(preimage) if secured else preimage
 
-T = TypeVar("T")
+        if preimage in skip:
+            continue
 
+        if obj[preimage] == b"" or obj[preimage] is None:
```
Comment (Bhargavasomu) [inline src/eth1spec/trie.py]: Why not
```
if obj[preimage] in (b"", None):
    ...
```

File: src/eth1spec/trie.py [hunk 4]
```diff
@@ -13,206 +13,182 @@
...
-    return nibbles, t
+    mapped: MutableMapping[Bytes, Node] = {}
+    skip: MutableMapping[Bytes, bool] = {}
 
+    for preimage in obj:
+        # "secure" tries hash keys once before construction
+        key = crypto.keccak256(preimage) if secured else preimage
 
-T = TypeVar("T")
+        if preimage in skip:
```
Comment (Bhargavasomu) [inline src/eth1spec/trie.py:105]: I prefer putting this line before calculating the key as a very tiny optimization. Let me know your thoughts.

File: src/eth1spec/trie.py [hunk 5]
```diff
@@ -13,206 +13,182 @@
...
-    nibbles = ""
-    odd_length = (buffer[0] >> 4) % 2 == 1  # sixth lowest bit
-    t = (buffer[0] >> 5) % 2 != 0  # fifth lowest bit
-    if odd_length:
-        nibbles += buffer[0:1].hex()[1]
-    for b in buffer[1:]:
-        nibbles += bytes([b]).hex()
-    return nibbles, t
+    mapped: MutableMapping[Bytes, Node] = {}
+    skip: MutableMapping[Bytes, bool] = {}
```
Comment (Bhargavasomu) [inline src/eth1spec/trie.py]: To reduce the load on the below for loop and for better readability, I propose moving the skip list/set generation part outside the for loop as follows.
```
skip: Set[Bytes] = {
    preimage
    for preimage in in obj
    if obj[preimage] in (b"", None)
}
```

File: src/eth1spec/trie.py (hunk 3 above)
Comment (Bhargavasomu) [reply src/eth1spec/trie.py]: Also this if block can be removed if the skip set generation part is done outside the for loop as commented above.

File: src/eth1spec/trie.py [hunk 6]
```diff
@@ -13,206 +13,182 @@
...
+Node = Union[Account, Bytes, Receipt, Uint, U256]
 
-def HP(x: Bytes, t: Union[bool, int]) -> bytearray:
+
+def nibble_list_to_compact(x: Bytes, terminal: bool) -> bytearray:
     """
-    Hex prefix encoding.
+    Compresses nibble-list into a standard byte array with a flag.
+
+    A nibble-list is a list of byte values no greater than `16`. The flag is
```
Comment (Bhargavasomu) [inline src/eth1spec/trie.py]: Shouldn't this be `15` instead of `16`? Please correct me if I am wrong. Thanks.

File: src/eth1spec/trie.py [hunk 7]
```diff
@@ -13,206 +13,182 @@
...
+    return crypto.keccak256(rlp.encode(root_node))
 
 
-def n(
-    J: Mapping[Bytes, Union[Bytes, Account, Receipt, Uint, U256]], i: U256
-) -> Bytes:
+def node_cap(obj: Mapping[Bytes, Node], i: Uint) -> rlp.RLP:
     """
-    Node composition function.
+    Internal nodes less than 32 bytes in length are represented by themselves
```
Comment (Bhargavasomu) [inline src/eth1spec/trie.py:143]: `Internal nodes whose rlp encoding is less than 32 bytes...`?

**Review by Bhargavasomu — APPROVED** [review]

File: src/eth1spec/trie.py (hunk 7 above)
Comment (lightclient) [reply src/eth1spec/trie.py:143]: meh

File: src/eth1spec/trie.py [hunk 8]
```diff
@@ -13,206 +13,181 @@
...
-    Returns
-    -------
-    flag : `int`
-        `t` encoded as a bit flag.
-    """
-    if t:
-        return 0b10
-    else:
-        return 0
+T = TypeVar("T")
```
Comment (SamWilsn) [inline src/eth1spec/trie.py]: Is this still used anywhere?

File: tests/helpers.py [hunk 9]
```diff
@@ -42,7 +50,13 @@ def hex2u256(x: str) -> U256:
     return U256(int(x, 16))
 
 
-def remove_hex_prefix(x: str) -> str:
+def has_hex_prefix(x: str) -> bool:
     if x.startswith("0x"):
+        return True
+    return False
```
Comment (SamWilsn) [inline tests/helpers.py]: Can we just do:

```suggestion
def has_hex_prefix(x: str) -> bool:
    return x.startswith("0x")
```

File: whitelist.txt [hunk 10]
```diff
@@ -48,6 +48,8 @@ ipow
 pos
 dest
 fromhex
+preimage
+substring
 
 elle
```
Comment (SamWilsn) [inline whitelist.txt]: I think this is safe to remove now.

**Review by SamWilsn — CHANGES_REQUESTED** [review]

File: src/eth1spec/trie.py (hunk 8 above)
Comment (lightclient) [reply src/eth1spec/trie.py]: doesn't look like it

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**